apps/web/src/app/(profile)/[username]/_components/dashboard/badges/_difficulty_badges.ts

```diff
@@ -0,0 +1,58 @@
+import { prisma } from '@repo/db';
+import type {AllBadgeObjs, BadgeFn} from "~/app/(profile)/[username]/_components/dashboard/_actions";
+import type {SolutionBadges} from "~/app/(profile)/[username]/_components/dashboard/badges/_shared_solutions_badges";
+import type {AotBadges} from "~/app/(profile)/[username]/_components/dashboard/badges/_advent_badges";
+export const DifficultyBadgeKeys = ['BEGINNER', 'EASY', 'MEDIUM', 'HARD', 'EXTREME'] as const;
+
+export type DifficultyBadges = typeof DifficultyBadgeKeys[number];
+export interface Difficulty {
+  Difficulty: DifficultyBadges;
+  TotalCompleted: number;
+}
+
+export const difficultyBadgesFn: BadgeFn =
+  async ({
+           userId,
+           badges,
+         }: {
+    userId: string;
+    badges: AllBadgeObjs;
+  }) => {
+    // DifficultyBadges
+    const difficulty: Difficulty[] | null = await DifficultyRetrieveData(userId);
+    return await DifficultyBadgesFn(badges, difficulty ?? []);
+  };
+
+export async function DifficultyRetrieveData(userId: string) {
+  const data: Difficulty[] | null =
+    await prisma.$queryRaw`SELECT Difficulty, COUNT(Id) as TotalCompleted FROM (SELECT DISTINCT Difficulty, Challenge.Id FROM Submission JOIN Challenge ON Submission.challengeId = Challenge.Id WHERE Submission.userId = ${userId} AND IsSuccessful = 1) unique_query GROUP BY Difficulty `;
```

Comment:
prisma syntax

Reply:
this might be fine actually. see if you can pull it off in prisma though

Reply:
Can you be more specific about "type to represent this or something"? There is a Difficulty[] | null type specified...

I only briefly tried to do this in prisma, really think the juice isn't worth the squeeze on this one.

Reply:
```suggestion
    await prisma.$queryRaw<Difficulty[]>`SELECT Difficulty, COUNT(Id) as TotalCompleted FROM (SELECT DISTINCT Difficulty, Challenge.Id FROM Submission JOIN Challenge ON Submission.challengeId = Challenge.Id WHERE Submission.userId = ${userId} AND IsSuccessful = 1) unique_query GROUP BY Difficulty `;
```

Doing this allows you to remove the type in data: type. Also you can remove null ( prisma does not returns null, it throws in case of an error, or returns an empty array ) so the full result should be:
```ts
    const data = await prisma.$queryRaw<Difficulty[]>`SELECT Difficulty, COUNT(Id) as TotalCompleted FROM (SELECT DISTINCT Difficulty, Challenge.Id FROM Submission JOIN Challenge ON Submission.challengeId = Challenge.Id WHERE Submission.userId = ${userId} AND IsSuccessful = 1) unique_query GROUP BY Difficulty `;
```

Reply:
got it, thanks typesafe. I've updated the type usage to the correct one.